elki-outlier/src/main/java/elki/outlier/isolation/IsolationForest.java

```diff
@@ -0,0 +1,366 @@
+   * argument is less than or equal to 1.
+   */
+  protected static double c(double n) {
+    if(n <= 1.0) {
+      return 0;
+    }
+    return 2 * (FastMath.log(n - 1) + 0.5772156649) - (2 * (n - 1) / n);
+  }
+
+  /**
+   * Returns distribution of scores.
+   */
+  // protected double[] distributionForInstance(V vector) {
+  protected double distributionForInstance(NumberVector vector) {
+    double avgPathLength = 0;
+    for(Tree m_tree : trees) {
+      avgPathLength += m_tree.pathLength(vector);
+    }
+    avgPathLength /= trees.size();
+
+    return FastMath.pow(2, -avgPathLength / c(subsampleSize));
+
+    // return scores;
+  }
+
+  /**
+   * Inner class for building and using an isolation tree.
+   */
+  static class Tree {
+
+    /**
+     * The size of the node
+     */
+    private int size;
+
+    /**
+     * The split attribute
+     */
+    private int a;
+
+    /**
+     * The split point
+     */
+    private double q;
+
+    /**
+     * The successors
+     */
+    private List<Tree> successors;
+
+    /**
+     * Constructs a tree from data.
+     *
+     * @param data Subsample from data
+     * @param r Random generator
+     * @param height
+     * @param maxHeight
+     */
+    protected Tree(ModifiableDBIDs data, Random r, int height, int maxHeight, Relation<? extends NumberVector> relation) {
+      // Set size of node
+      size = data.size();
+
+      // Stop splitting if necessary
+      if((size <= 1) || (height == maxHeight)) {
+        return;
+      }
+
+      // Compute mins and maxs and eligible attributes
+      final int dim = RelationUtil.dimensionality(relation);
+      double[][] minmax;
+      {
+        double[] mins = new double[dim], maxs = new double[dim];
+        for(int i = 0; i < dim; i++) {
+          mins[i] = Double.MAX_VALUE;
+          maxs[i] = -Double.MAX_VALUE;
+        }
+        for(DBIDIter iditer = data.iter(); iditer.valid(); iditer.advance()) {
+          final NumberVector o = relation.get(iditer);
+          for(int d = 0; d < dim; d++) {
+            final double v = o.doubleValue(d);
+            mins[d] = Math.min(v, mins[d]);
+            maxs[d] = Math.max(v, maxs[d]);
+          }
+        }
+        minmax = new double[][] { mins, maxs };
+      }
+
+      ArrayList<Integer> al = new ArrayList<Integer>();
+      for(int j = 0; j < dim; j++) {
+        if(minmax[0][j] < minmax[1][j]) {
+          al.add(j);
+        }
+      }
+
+      // Check whether any eligible attributes have been found
+      if(al.size() == 0) {
+        return;
+      }
+      else {
+        // Randomly pick an attribute and split point
+        a = al.get(r.nextInt(al.size()));
+        q = (r.nextDouble() * (minmax[1][a] - minmax[0][a])) + minmax[0][a];
+
+        // Create sub trees
+        successors = new ArrayList<>(2);
+        for(int i = 0; i < 2; i++) {
+          ArrayModifiableDBIDs tmpData = DBIDUtil.newArray();
```

Comment:
Allocating the array with `data.size()` is likely a bit faster, and the temporary additional memory usage should be okay.